File: python_ui/utils/config_manager.py
"""
Configuration Manager
Handles loading and saving config.json
"""

import json
import os
from pathlib import Path

class ConfigManager:
    def __init__(self, config_path=None):
        if config_path is None:
            # Default to parent directory config.json
            base_dir = Path(__file__).parent.parent.parent
            self.config_path = base_dir / 'config.json'
        else:
            self.config_path = Path(config_path)
        
        self.config = {}
        self.load()
    
    def load(self):
        """Load configuration from file"""
        if not self.config_path.exists():
            # Create default config
            self.config = {
                'adspowerApiKey': '',
                'adspowerProfileId': '',
                'facebookPageId': '',
                'metaBusinessUrl': 'https://business.facebook.com/latest/home',
                'uploadFolderPath': '',
                'debugFolderPath': '',
                'headless': False,
                'debugMode': True,
                'maxRetries': 3,
                'uploadTimeoutSeconds': 300
            }
            self.save()
        else:
            try:
                with open(self.config_path, 'r', encoding='utf-8') as f:
                    self.config = json.load(f)
            except Exception as e:
                print(f"Error loading config: {e}")
                self.config = {}
    
    def save(self):
        """Save configuration to file"""
        try:
            with open(self.config_path, 'w', encoding='utf-8') as f:
                json.dump(self.config, f, indent=2, ensure_ascii=False)
            return True
        except Exception as e:
            print(f"Error saving config: {e}")
            return False
    
    def get(self, key, default=None):
        """Get a configuration value"""
        return self.config.get(key, default)
    
    def set(self, key, value):
        """Set a configuration value"""
        self.config[key] = value
    
    def get_all(self):
        """Get all configuration"""
        return self.config.copy()
    
    def update(self, updates):
        """Update multiple configuration values"""
        self.config.update(updates)

File: utils/configManager.js
/**
 * Configuration Manager
 * Handles configuration file management using Electron's userData directory
 * Ensures config is stored in a writable location when app is installed
 */

const { app } = require('electron');
const fs = require('fs');
const path = require('path');

/**
 * Get the config file path in userData directory
 * @returns {string} Full path to config.json
 */
function getConfigPath() {
    const userDataPath = app.getPath('userData');
    return path.join(userDataPath, 'config.json');
}

/**
 * Get default configuration values
 * @returns {Object} Default config object
 */
function getDefaultConfig() {
    return {
        adspowerApiKey: '',
        adspowerProfileId: '',
        facebookPageId: '',
        uploadFolderPath: '',
        debugMode: true,
        maxRetries: 3,
        uploadTimeoutSeconds: 300
    };
}

/**
 * Load configuration from userData
 * Creates default config if file doesn't exist
 * @returns {Object} Configuration object
 */
function loadConfig() {
    try {
        const configPath = getConfigPath();

        // Create config file with defaults if it doesn't exist
        if (!fs.existsSync(configPath)) {
            console.log('[CONFIG] Config file not found, creating default config...');
            const defaultConfig = getDefaultConfig();
            saveConfig(defaultConfig);
            return defaultConfig;
        }

        // Read and parse existing config
        const configData = fs.readFileSync(configPath, 'utf8');
        const config = JSON.parse(configData);
        
        console.log(`[CONFIG] Configuration loaded from: ${configPath}`);
        return config;
    } catch (error) {
        console.error(`[CONFIG] Error loading config: ${error.message}`);
        console.log('[CONFIG] Returning default config...');
        return getDefaultConfig();
    }
}

/**
 * Save configuration to userData
 * @param {Object} config - Configuration object to save
 * @returns {boolean} Success status
 */
function saveConfig(config) {
    try {
        const configPath = getConfigPath();
        const userDataPath = path.dirname(configPath);

        // Ensure userData directory exists
        if (!fs.existsSync(userDataPath)) {
            fs.mkdirSync(userDataPath, { recursive: true });
        }

        // Write config file
        fs.writeFileSync(configPath, JSON.stringify(config, null, 2), 'utf8');
        console.log(`[CONFIG] Configuration saved to: ${configPath}`);
        return true;
    } catch (error) {
        console.error(`[CONFIG] Error saving config: ${error.message}`);
        return false;
    }
}

module.exports = {
    getConfigPath,
    loadConfig,
    saveConfig,
    getDefaultConfig
};
